File: Train/training_data_patch.m
function training_data_patch(training_bf_patch_dir, mask_dir, aug_process, sub_num, aug_num, symmetry_mode)

%
% Description:
%    Patching training data into 64 x 64 x 64 matrices
%   
% Inputs:
%   training_bf_patch_dir : directory of training data before patch
%                           inputs  - local field data (multiphs)
%                           outputs - cosmos susceptibility data (multicos)
%   mask_dir : director of mask data
%   aug_process : data augmentation type - 1 for QSMnet, 2 for QSMnet+ 
%   sub_num : subject number (max 7)
%   aug_num : augmentation number (max 5)
%   symmetry_mode : data augmentation by inverting the sign of trainig datset
%                   if symmetry_mode == 1, training data size become double
%
% Outputs:
%   save training_data_patch_64.mat 
%
% Copyright @ Woojin Jung
% Laboratory for Imaging Science and Technology
% Seoul National University
% email: wjjung93@snu.ac.kr
%


load(training_bf_patch_dir);
load(mask_dir);

scaling_factor = 4;
patch_stride = [7, 8, 6];
matrix_cut = [172, 176, 159];

PS = 64; %patch size 64 - 172, 176, 159 
str_x = (matrix_cut(1)-PS)/(patch_stride(1)-1);
str_y = (matrix_cut(2)-PS)/(patch_stride(2)-1);
str_z = (matrix_cut(3)-PS)/(patch_stride(3)-1);


[yy, xx] = meshgrid(1:patch_stride(2),1:patch_stride(1));
xx = repmat(xx,[1,1,patch_stride(3)]);
yy = repmat(yy,[1,1,patch_stride(3)]);
for kk=1:patch_stride(3)
    zz(:,:,kk) = ones(patch_stride(1),patch_stride(2))*kk;
end
xx=xx(:);
yy=yy(:);
zz=zz(:);
tt=length(xx);

data_input = single(zeros(PS,PS,PS,sub_num,tt*(5+aug_num)));
data_label = single(zeros(PS,PS,PS,sub_num,tt*(5+aug_num)));
mask_input = single(zeros(PS,PS,PS,sub_num,tt*(5+aug_num)));

matrix_size=[176, 176, 160];
voxel_size=[1,1,1];
B0_dir = [0, 0, 1];
D = single(dipole_kernel(matrix_size,voxel_size,B0_dir));
input_voxel = [];
output_voxel = [];

for ii=1:sub_num
    eval(sprintf('data1 = multicos%d;',ii));
    eval(sprintf('data2 = multiphs%d;',ii));
    eval(sprintf('maskt = mask%d;',ii));
    % Original training dataset
    for aug = 1:5
        data_L = data1(3:174,:,1:159,aug);
        data_I = data2(3:174,:,1:159,aug);
        mask_L = maskt(3:174,:,1:159,aug);
        
        % Process for data normalization
        cc = data_L;
        pp = data_I;
        mm = mask_L;
        cc(mm==0)=0;
        cc=cc(:);
        cc(cc==0)=[];
        pp(mm==0)=0;
        pp=pp(:);
        pp(pp==0)=[];
        input_voxel = cat(1,input_voxel,pp);
        output_voxel = cat(1,output_voxel,cc);

        for jj=1:tt;
            data_input(1,:,:,:,ii,tt*(aug-1)+jj) = data_I(str_x*xx(jj)-(str_x-1):str_x*xx(jj)+(PS-str_x),...
                str_y*yy(jj)-(str_y-1):str_y*yy(jj)+(PS-str_y),str_z*zz(jj)-(str_z-1):str_z*zz(jj)+(PS-str_z));
            data_label(1,:,:,:,ii,tt*(aug-1)+jj) = data_L(str_x*xx(jj)-(str_x-1):str_x*xx(jj)+(PS-str_x),...
                str_y*yy(jj)-(str_y-1):str_y*yy(jj)+(PS-str_y),str_z*zz(jj)-(str_z-1):str_z*zz(jj)+(PS-str_z));
            mask_input(1,:,:,:,ii,tt*(aug-1)+jj) = mask_L(str_x*xx(jj)-(str_x-1):str_x*xx(jj)+(PS-str_x),...
                str_y*yy(jj)-(str_y-1):str_y*yy(jj)+(PS-str_y),str_z*zz(jj)-(str_z-1):str_z*zz(jj)+(PS-str_z));
        end
    end
    % Data augmentation by dipole convolution
    if aug_num < 1 
        break;
    else
        for aug = 6:aug_num+5
            data_L = data1(:,:,:,aug);
            if aug_process == 1 % QSMnet (no scaling factor applied)
                data_L = data1(3:174,:,1:159,aug);
                net_name = 'QSMnet';
            elseif aug_process == 2 % QSMnet+
                data_L(data_L>0) = data_L(data_L>0)*scaling_factor;
                net_name = 'QSMnet+';
            end
            data_I = ifftn(fftn(data_L).*D);
            
            data_L = data_L(3:174,:,1:159,aug);
            data_I = data_I(3:174,:,1:159,aug);
            mask_L = maskt(3:174,:,1:159,aug);
            
            % Process for data normalization
            cc = data_L;
            pp = data_I;
            mm = mask_L;
            cc(mm==0)=0;
            cc=cc(:);
            cc(cc==0)=[];
            pp(mm==0)=0;
            pp=pp(:);
            pp(pp==0)=[];
            input_voxel = cat(1,input_voxel,pp);
            output_voxel = cat(1,output_voxel,cc); 
            
            for jj=1:tt;
                data_input(1,:,:,:,ii,tt*(aug-6)+jj) = data_I(str_x*xx(jj)-(str_x-1):str_x*xx(jj)+(PS-str_x),...
                    str_y*yy(jj)-(str_y-1):str_y*yy(jj)+(PS-str_y),str_z*zz(jj)-(str_z-1):str_z*zz(jj)+(PS-str_z));
                data_label(1,:,:,:,ii,tt*(aug-6)+jj) = data_L(str_x*xx(jj)-(str_x-1):str_x*xx(jj)+(PS-str_x),...
                    str_y*yy(jj)-(str_y-1):str_y*yy(jj)+(PS-str_y),str_z*zz(jj)-(str_z-1):str_z*zz(jj)+(PS-str_z));
                mask_input(1,:,:,:,ii,tt*(aug-6)+jj) = mask_L(str_x*xx(jj)-(str_x-1):str_x*xx(jj)+(PS-str_x),...
                    str_y*yy(jj)-(str_y-1):str_y*yy(jj)+(PS-str_y),str_z*zz(jj)-(str_z-1):str_z*zz(jj)+(PS-str_z));
            end
        end
    end
end

data_input = data_input(1,:,:,:,:);
data_label = data_label(1,:,:,:,:);
mask_input = mask_input(1,:,:,:,:);

data_input = single(permute(data_input,[1 4 3 2 5])); 
data_label = single(permute(data_label,[1 4 3 2 5]));
mask_input = single(permute(mask_input,[1 4 3 2 5]));

if symmetry_mode == 1
    sym_name = 'sym';
    data_input = cat(4, data_input, -data_input);
    data_label = cat(4, data_label, -data_label);
    mask_input = cat(4, mask_input, mask_input);
    input_voxel = cat(1,input_voxel, -input_voxel);
    output_voxel = cat(1,output_voxel, -output_voxel);
else
    sym_name = 'asym';
end
save_data_name = ['training_data_patch_64_',net_name,'_',sym_name,'.mat'];
save_norm_name = ['norm_factor_',net_name,'_',sym_name,'.mat'];

input_mean = mean(input_voxel);
input_std = std(input_voxel);
label_mean = mean(output_voxel);
label_std = std(output_voxel);

eval(sprintf('save %s data_input data_label mask_input -v7.3',save_data_name));
eval(sprintf('save %s input_mean input_std label_mean label_std -v7.3',save_norm_name));

File: Code/training_data_patch.py
import scipy.io
import numpy as np
import h5py
import time
import os
import sys
#
# Description:
# Patching training data
#
# Inputs:
#   training_bf_patch_dir : directory of training data before patch
#                         : inputs - local field data (phs_tissue)
#                         : outputs - cosmos susceptibility data (chi_cosmos)
#   mask_dir : directory of mask data (mask)
#   PS : patch size
#   sub_num : subject number
#   aug_num : augmentation number
#   patch_num : the number of patches [x,y,z] per subject
#
# Outputs:
#   save training_data_patch_64.mat
#
# Copyright @ Woojin Jung
# Laboratory for Imaging Science and Technology
# Seoul National University
# email: wjjung93@snu.ac.kr
#

'''
File Path
'''
FILE_PATH_INPUT = '../Data/Train/Input/train_input'
FILE_PATH_MASK = '../Data/Train/Input/mask'
FILE_PATH_LABEL = '../Data/Train/Label/train_label'
start_time = time.time()

'''
Constant Variables
'''
PS = 32  # Patch size
net_name = 'QSMnet'
sub_num = 1  # number of subjects
dir_num = 1  # number of directions
patch_num = [5, 5, 5]  # Order of Dimensions: [x, y, z]

'''
Code Start
'''

# Create Result File
result_file = h5py.File(
    '../Data/Train/Training_data_patch/training_data_patch_'+ str(net_name) + '_' + str(PS) + '.hdf5', 'w')

# Patch the input & mask file ----------------------------------------------------------------


print("####patching input####")
patches = []
patches_mask = []
for dataset_num in range(1, sub_num + 1):
    field = scipy.io.loadmat(FILE_PATH_INPUT + str (dataset_num) + '.mat')
    mask = scipy.io.loadmat(FILE_PATH_MASK + str(dataset_num) + '.mat')
    matrix_size = np.shape(field['phs_tissue'])
    strides = [(matrix_size[i] - PS) // (patch_num[i] - 1) for i in range(3)]
    
    if np.size(matrix_size) == 3:
        field['phs_tissue'] = np.expand_dims(field['phs_tissue'], axis = 3)
        matrix_size = np.append(matrix_size, [1], axis=0)
        mask['mask'] = np.expand_dims(mask['mask'], axis = 3)
    
    if matrix_size[3] < dir_num:
        sys.exit("dir_num is bigger than data size!")
        
    for idx in range(dir_num):
        for i in range(patch_num[0]):
            for j in range(patch_num[1]):
                for k in range(patch_num[2]):
                    patches.append(field['phs_tissue'][
                                   i * strides[0]:i * strides[0] + PS,
                                   j * strides[1]:j * strides[1] + PS,
                                   k * strides[2]:k * strides[2] + PS,
                                   idx])
                    patches_mask.append(mask['mask'][
                                        i * strides[0]:i * strides[0] + PS,
                                        j * strides[1]:j * strides[1] + PS,
                                        k * strides[2]:k * strides[2] + PS,
                                        idx])
print("Done!")

patches = np.array(patches, dtype='float32', copy=False)
patches_mask = np.array(patches_mask, dtype='float32', copy=False)

patches = np.expand_dims(patches,axis=4)
patches_mask = np.expand_dims(patches_mask,axis=4)
print("Final input data size : " + str(np.shape(patches)))

input_mean = np.mean(patches[patches_mask > 0])
input_std = np.std(patches[patches_mask > 0])

#patches = (patches - input_mean) / input_std

result_file.create_dataset('temp_i', data=patches)
result_file.create_dataset('temp_m', data=patches_mask)

del patches

# Patch the label file --------------------------------------------------------------------

patches = []
print("####patching label####")
for dataset_num in range(1, sub_num + 1):
    susc = scipy.io.loadmat(FILE_PATH_LABEL + str (dataset_num) + '.mat')
    matrix_size = np.shape(susc['chi_cosmos'])
    strides = [(matrix_size[i] - PS) // (patch_num[i] - 1) for i in range(3)]
    
    if np.size(matrix_size) == 3:
        susc['chi_cosmos'] = np.expand_dims(susc['chi_cosmos'], axis = 3)
        matrix_size = np.append(matrix_size, [1], axis=0)
        
    if matrix_size[3] < dir_num:
        sys.exit("dir_num is bigger than data size!")
        
    for idx in range(dir_num):
        for i in range(patch_num[0]):
            for j in range(patch_num[1]):
                for k in range(patch_num[2]):
                    patches.append(susc['chi_cosmos'][
                                   i * strides[0]:i * strides[0] + PS,
                                   j * strides[1]:j * strides[1] + PS,
                                   k * strides[2]:k * strides[2] + PS,
                                   idx])
print("Done!")

patches = np.array(patches, dtype='float32', copy=False)
patches = np.expand_dims(patches,axis=4)    
print("Final label data size : " + str(np.shape(patches)))

label_mean = np.mean(patches[patches_mask > 0])
label_std = np.std(patches[patches_mask > 0])

#patches = (patches - label_mean) / label_std

result_file.create_dataset('temp_l', data=patches)

del patches
del patches_mask
result_file.close()

save_path = '../Checkpoints/' + str(net_name) + '_'+ str(PS) + '/'
if not os.path.exists(save_path):
    os.makedirs(save_path)
scipy.io.savemat(save_path + 'norm_factor.mat',
                 mdict={'input_mean': input_mean, 'input_std': input_std,
                        'label_mean': label_mean, 'label_std': label_std})
print("Total Time: {:.3f}".format(time.time() - start_time))
